File: Graphs/Medium/SingleCycleCheck.py
def hasSingleCycle(array):
    #Name: Single Cycle Check
	#Category and difficulty: Graphs, Medium
    #time: O(n), where n is the length of the array
	#space: O(1)
    index = 0
    length = len(array)
    count = 0
    mark = -123123123123123
    startMark = -999999

    #Only continue while not repeating elements in array
    while(count < length):
        #When a zero is encountered, we will never move anywhere
        if(array[index] == 0):
            return False
        #If reached starting number before going through entire array, more than one cycle
        if(array[index] == startMark and count < length):
            return False
        
        count += 1
        sum = index + array[index]

        #Mark number as simply visited or starting index
        if(index == 0):
            array[index] = startMark
        else:
            array[index] = mark

        #Increase index 
        if(sum < 0 or sum >= length):
            index = sum % length
        else:
            index = sum

        #Check if cycling in array
        if(count > length+1 or array[index] == mark):
            return False
            
    return True
    pass

File: Graphs/Medium/SingleCycleCheck.cpp
using namespace std;

bool hasSingleCycle(vector<int> array) {
// #Name: Single Cycle Check
// #Category and difficulty: Graphs, Medium
// #time: O(n), where n is the length of the array
// #space: O(1)
  int count = 0;
  int visited = -999999;
  int start = 999999;
  int index = 0;
  
  while(count < array.size())
    {
      //Edge case
      if(array[index] == 0) {return false;}

      //check if reached start or previously visited node before checking all nodes 
      if((array[index] == start || array[index] == visited) && count <= array.size())
      {
        return false;
      }

      //get jump value
      int jump = array[index];

      //Mark current node as either start or visited
      if(index == 0)
      {
        array[index] = start;
      }
      else
      {
        array[index] = visited;
      }
      count++;

      
      index = (index + jump) % (int)array.size();
      if(index < 0)
      {
        index = index + array.size();
      }
      
      if(count > array.size() || array[index] == visited)
      {
        return false;
      }
    }
  
  return true;
}
